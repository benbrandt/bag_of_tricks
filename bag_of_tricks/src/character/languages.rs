use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use super::Character;

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Auran,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub(crate) fn gen(rng: &mut impl Rng, character: &Character, amount: usize) -> Vec<Self> {
        Language::iter()
            .filter(|l| !character.languages.contains(l))
            .choose_multiple(rng, amount)
    }
}

/// Trait for describing what languages are provided by a given object.
pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> usize {
        0
    }
}
