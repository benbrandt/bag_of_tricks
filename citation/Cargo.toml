[package]
name = "citation"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1", features = ["derive"] }
strum = { version = "0.20", features = ["derive"] }
