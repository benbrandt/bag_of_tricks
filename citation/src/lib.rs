#![deny(clippy::all)]
#![warn(clippy::pedantic)]
use std::{
    collections::{HashMap, HashSet},
    fmt,
};

use serde::{Deserialize, Serialize};
use strum::Display;

/// Titles of the available D&D Books.
#[derive(Clone, Copy, Debug, Deserialize, Display, Eq, Hash, PartialEq, Serialize)]
pub enum Book {
    #[strum(serialize = "PHB")]
    Phb,
    #[strum(serialize = "SCAG")]
    Scag,
    #[strum(serialize = "VGTM")]
    Vgtm,
}

/// Book and page number for citations.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Citation(pub Book, pub u16);

impl fmt::Display for Citation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} p{}", self.0, self.1)
    }
}

/// List of multiple citations.
#[derive(Deserialize, Serialize)]
pub struct CitationList(pub Vec<Citation>);

/// Displays multiple citations from the same book together.
impl fmt::Display for CitationList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut citations = HashMap::new();
        for c in &self.0 {
            citations
                .entry(c.0)
                .or_insert_with(HashSet::new)
                .insert(c.1);
        }
        let mut books = citations
            .iter()
            .map(|(book, page_nums)| {
                let mut pages = page_nums
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<_>>();
                pages.sort();
                format!("{} p{}", book, pages.join(","))
            })
            .collect::<Vec<_>>();
        books.sort();
        write!(f, "{}", books.join(" "))
    }
}

/// Trait for any entity/object in need of citation.
///
/// Makes it easer for users to find more information in the source books.
pub trait Citations {
    /// Return list of citations for the object in question
    fn citations(&self) -> CitationList;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_book_display() {
        assert_eq!(format!("{}", Book::Phb), "PHB");
    }

    #[test]
    fn test_citation_display() {
        let citation = Citation(Book::Phb, 123);
        assert_eq!(format!("{}", citation), "PHB p123");
    }

    #[test]
    fn test_citations_display() {
        let citations = CitationList(vec![
            Citation(Book::Phb, 123),
            Citation(Book::Phb, 125),
            Citation(Book::Vgtm, 125),
        ]);
        assert_eq!(format!("{}", citations), "PHB p123,125 VGTM p125");
    }
}
