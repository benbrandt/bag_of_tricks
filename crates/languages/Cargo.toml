[package]
name = "languages"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
deities = { path = "../deities" }
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.22", features = ["derive"] }
