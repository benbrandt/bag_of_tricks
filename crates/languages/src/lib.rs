use deities::{Pantheon, PantheonWeight, Pantheons};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

#[derive(Clone, Copy, Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum LanguageType {
    Exotic,
    Standard,
}

impl LanguageType {
    fn weight(self) -> f64 {
        match self {
            Self::Exotic => 1.0,
            Self::Standard => 2.0,
        }
    }
}

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gith,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub fn gen(
        rng: &mut impl Rng,
        current_languages: &[Language],
        (amount, language_type): (usize, Option<LanguageType>),
    ) -> Vec<Self> {
        language_type
            .map_or_else(
                || Self::iter().collect::<Vec<Self>>(),
                |lt| {
                    let filtered = Self::iter()
                        .filter(|l| l.language_type() == lt)
                        .collect::<Vec<Self>>();
                    if filtered.is_empty() {
                        Self::iter().collect::<Vec<Self>>()
                    } else {
                        filtered
                    }
                },
            )
            .iter()
            .filter(|l| !current_languages.contains(l))
            .copied()
            .collect::<Vec<_>>()
            .choose_multiple_weighted(rng, amount, |l| l.language_type().weight())
            .unwrap()
            .copied()
            .collect()
    }

    fn language_type(self) -> LanguageType {
        match self {
            Self::Common
            | Self::Dwarvish
            | Self::Elvish
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Orc => LanguageType::Standard,
            Self::Abyssal
            | Self::Celestial
            | Self::DeepSpeech
            | Self::Draconic
            | Self::Gith
            | Self::Infernal
            | Self::Primordial
            | Self::Sylvan
            | Self::Undercommon => LanguageType::Exotic,
        }
    }
}

impl Pantheons for Language {
    fn addl_pantheons(&self) -> Vec<(Pantheon, PantheonWeight)> {
        (match self {
            Language::Abyssal
            | Language::Celestial
            | Language::DeepSpeech
            | Language::Gith
            | Language::Infernal
            | Language::Primordial
            | Language::Sylvan => vec![],
            Language::Common => vec![Pantheon::ForgottenRealms],
            Language::Draconic => vec![Pantheon::Dragon, Pantheon::Kobold, Pantheon::Lizardfolk],
            Language::Dwarvish => vec![Pantheon::Dwarven],
            Language::Elvish => vec![Pantheon::Elven],
            Language::Giant => vec![Pantheon::Giant],
            Language::Gnomish => vec![Pantheon::Gnomish],
            Language::Goblin => vec![Pantheon::Bugbear, Pantheon::Goblin],
            Language::Halfling => vec![Pantheon::Halfling],
            Language::Orc => vec![Pantheon::Orc],
            Language::Undercommon => vec![Pantheon::Drow, Pantheon::Duergar],
        })
        .into_iter()
        .map(|p| (p, PantheonWeight::Possible))
        .collect::<Vec<_>>()
    }
}

/// Trait for describing what languages are provided by a given object.
pub trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (0, None)
    }
}
