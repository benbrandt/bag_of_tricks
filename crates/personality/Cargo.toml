[package]
name = "personality"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
alignment = { path = "../alignment" }
serde = { version = "1", features = ["derive"] }
strum = { version = "0.25", features = ["derive"] }
