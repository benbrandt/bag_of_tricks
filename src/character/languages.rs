use serde::{Deserialize, Serialize};
use strum_macros::{Display, EnumIter};

#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Does this provide additional language options?
    fn addl_languages(&self) -> usize {
        0
    }
}
