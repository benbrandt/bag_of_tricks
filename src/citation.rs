use serde::{Deserialize, Serialize};
use std::{
    collections::{HashMap, HashSet},
    fmt,
};
use strum_macros::Display;

/// Titles of the available D&D Books.
#[derive(Clone, Copy, Debug, Deserialize, Display, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Book {
    #[strum(serialize = "PHB")]
    Phb,
    #[strum(serialize = "VGTM")]
    Vgtm,
}

/// Book and page number for citations.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub(crate) struct Citation(pub(crate) Book, pub(crate) u16);

impl fmt::Display for Citation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} p{}", self.0, self.1)
    }
}

/// List of multiple citations.
#[derive(Deserialize, Serialize)]
pub(crate) struct CitationList(pub(crate) Vec<Citation>);

/// Displays multiple citations from the same book together.
impl fmt::Display for CitationList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut citations = HashMap::new();
        for c in &self.0 {
            citations
                .entry(c.0)
                .or_insert_with(HashSet::new)
                .insert(c.1);
        }
        for (book, page_nums) in citations {
            let mut pages = page_nums
                .iter()
                .map(ToString::to_string)
                .collect::<Vec<_>>();
            pages.sort();
            write!(f, "{} p{}", book, pages.join(","))?;
        }
        write!(f, "")
    }
}

/// Trait for any entity/object in need of citation.
///
/// Makes it easer for users to find more information in the source books.
pub(crate) trait Citations {
    /// Return list of citations for the object in question
    fn citations(&self) -> CitationList;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_book_display() {
        assert_eq!(format!("{}", Book::Phb), "PHB");
    }

    #[test]
    fn test_citation_display() {
        let citation = Citation(Book::Phb, 123);
        assert_eq!(format!("{}", citation), "PHB p123");
    }

    #[test]
    fn test_citations_display() {
        let citations = CitationList(vec![Citation(Book::Phb, 123), Citation(Book::Phb, 125)]);
        assert_eq!(format!("{}", citations), "PHB p123,125");
    }
}
