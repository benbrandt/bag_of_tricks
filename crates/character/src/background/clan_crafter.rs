use std::fmt;

use citation::{Book, Citation, CitationList, Citations};
use languages::{Language, Languages};
use personality::{Influence, PersonalityOptions};
use rand::Rng;
use serde::{Deserialize, Serialize};

use crate::{
    ability::Skill,
    backstory::Backstory,
    equipment::{
        adventuring_gear::{Gear, OtherGear},
        currency::Coin,
        Equipment, EquipmentOption, StartingEquipment,
    },
    features::{Feature, Features},
    proficiencies::{Proficiencies, Proficiency, ProficiencyOption},
    Character,
};

use super::{
    guild_artisan::{BONDS, FLAWS, IDEALS, TRAITS},
    Background,
};

const SKILLS: &[Skill] = &[Skill::History, Skill::Insight];

#[derive(Deserialize, Serialize)]
pub(crate) struct ClanCrafter;

#[typetag::serde]
impl Background for ClanCrafter {
    fn gen(_: &mut impl Rng, _: &Character) -> Box<dyn Background> {
        Box::new(Self)
    }

    fn skills() -> Vec<Skill> {
        SKILLS.to_vec()
    }
}

impl Backstory for ClanCrafter {}

impl Citations for ClanCrafter {
    fn citations(&self) -> CitationList {
        CitationList(vec![Citation(Book::Scag, 145)])
    }
}

impl Features for ClanCrafter {
    fn features(&self) -> Vec<Feature> {
        // As well respected as clan crafters are among outsiders, no one esteems them quite so highly as dwarves do. You always have free room and board in any place where shield dwarves or gold dwarves dwell, and the individuals in such a settlement might vie among themselves to determine who can offer you (and possibly your compatriots) the finest accommodations and assistance.
        vec![Feature {
            title: "Respect of the Stout Folk",
            citation: Citation(Book::Scag, 145),
        }]
    }
}

impl Languages for ClanCrafter {
    fn languages(&self) -> Vec<Language> {
        vec![Language::Dwarvish]
    }
}

impl PersonalityOptions for ClanCrafter {
    fn bonds(&self) -> Vec<String> {
        BONDS
            .iter()
            .map(|&s| s.to_string().replace("guild", "clan"))
            .collect()
    }

    fn flaws(&self) -> Vec<String> {
        FLAWS
            .iter()
            .map(|&s| s.to_string().replace("guild", "clan"))
            .collect()
    }

    fn ideals(&self) -> Vec<(String, Influence)> {
        IDEALS
            .iter()
            .map(|&(s, i)| (s.to_string().replace("guild", "clan"), i))
            .collect()
    }

    fn traits(&self) -> Vec<String> {
        TRAITS
            .iter()
            .map(|&s| s.to_string().replace("guild", "clan"))
            .collect()
    }
}

impl Proficiencies for ClanCrafter {
    fn proficiencies(&self) -> Vec<Proficiency> {
        SKILLS.iter().map(|&s| Proficiency::Skill(s)).collect()
    }

    fn addl_proficiencies(&self) -> Vec<ProficiencyOption> {
        vec![ProficiencyOption::ArtisansTools]
    }
}

impl StartingEquipment for ClanCrafter {
    // TODO: supposed to be 5gp + gem worth 10gp
    fn coins(&self) -> (Coin, u8) {
        (Coin::Gold, 15)
    }

    fn equipment(&self) -> Vec<Equipment> {
        vec![
            Equipment::Other("a maker's mark chisel used to mark your handiwork with the symbol of the clan of crafters you learned your skill from".into()),
            Equipment::Gear(Gear::Other(OtherGear::ClothesTravelers)),
            Equipment::Gear(Gear::Other(OtherGear::Pouch)),
        ]
    }

    fn addl_equipment(&self) -> Vec<EquipmentOption> {
        vec![EquipmentOption::ArtisansTools]
    }
}

impl fmt::Display for ClanCrafter {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Clan Crafter")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = ClanCrafter::gen(&mut rng, &Character::default());
        insta::assert_yaml_snapshot!(background);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = ClanCrafter::gen(&mut rng, &Character::default());
        insta::assert_display_snapshot!(background);
    }

    #[test]
    fn test_bonds() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = ClanCrafter::gen(&mut rng, &Character::default());
        insta::assert_yaml_snapshot!(background.bonds());
    }

    #[test]
    fn test_flaws() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = ClanCrafter::gen(&mut rng, &Character::default());
        insta::assert_yaml_snapshot!(background.flaws());
    }

    #[test]
    fn test_ideals() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = ClanCrafter::gen(&mut rng, &Character::default());
        insta::assert_yaml_snapshot!(background.ideals());
    }

    #[test]
    fn test_traits() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = ClanCrafter::gen(&mut rng, &Character::default());
        insta::assert_yaml_snapshot!(background.traits());
    }

    #[test]
    fn test_skills() {
        insta::assert_yaml_snapshot!(ClanCrafter::skills());
    }

    #[test]
    fn test_snapshot_citations() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.citations());
    }

    #[test]
    fn test_snapshot_features() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.features());
    }

    #[test]
    fn test_snapshot_languages() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.languages());
    }

    #[test]
    fn test_snapshot_proficiences() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.proficiencies());
    }

    #[test]
    fn test_snapshot_addl_proficiences() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.addl_proficiencies());
    }

    #[test]
    fn test_snapshot_coins() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.coins());
    }

    #[test]
    fn test_snapshot_equipment() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.equipment());
    }

    #[test]
    fn test_snapshot_addl_equipment() {
        let background = ClanCrafter;
        insta::assert_yaml_snapshot!(background.addl_equipment());
    }
}
