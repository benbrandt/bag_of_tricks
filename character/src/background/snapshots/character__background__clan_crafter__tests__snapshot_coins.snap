---
source: character/src/background/clan_crafter.rs
expression: background.coins()

---
- Gold
- 15

